### PR TITLE
Add ZeroOp as neutral elements of addition, make IdentityOp neutral element of composition

### PR DESCRIPTION
**Pull request by fzimmermann89**
Add ZeroOp (returns zeros)
Makes ZeroOp neutral element of OperatorAddion, IdentityOp of composition, i.e.

Op + zeroop = op
op@identityop = op

### CODE REVIEW

**Comment by fzimmermann89** [comment]
Curently in this pr, ZeroOp by default returns zeros of the same shape and type as the input. By setting the flag to false  it can be changed to instead return a scalar 0 tensors in forward and adjoint

I am considering changing the default of the ZeroOp to always return a scalar 0 tensor (which can broadcast/type promote to domain and range)

This is the most useful form of the operator ( as a neutral element of addition), for example in the matrix operator. On its own, it does not pass the adjointess test, but ZeroOp+any linearoperator does. 
I cannot think of a use case where retuning a scalar tensor does not work.

**Comment by fzimmermann89** [comment]
I am considering changing the default of the ZeroOp to always return a scalar integer 0 in forward and adjoint.
This is the most useful form of the operator ( as a neutral element of addition), for example in the matrix operator. On its own, it does not pass the adjointess test, but ZeroOp+any linearoperator does. 
I cannot think of a use case where retuning a scalar tensor does not work. 
And whenever somebody needs a zeroop, they actually need this broadcasting form..

**Comment by fzimmermann89** [comment]
@ckolbPTB @schuenke

**Comment by ckolbPTB** [comment]
> I am considering changing the default of the ZeroOp to always return a scalar integer 0 in forward and adjoint.

sounds reasonable to me

File: src/mrpro/operators/LinearOperator.py [hunk 1]
```diff
@@ -145,46 +148,93 @@ def __matmul__(
 
         Returns lambda x: self(other(x))
         """
-        if isinstance(other, LinearOperator):
+        if isinstance(other, mrpro.operators.IdentityOp):
```
Comment (ckolbPTB) [inline src/mrpro/operators/LinearOperator.py:206]: Any special reasons for why if is used for the first and if-elif for the rest?

File: src/mrpro/operators/LinearOperator.py [hunk 2]
```diff
@@ -145,46 +148,93 @@ def __matmul__(
...
+            return other
+        elif isinstance(other, LinearOperator):
             # LinearOperator@LinearOperator is linear
             return LinearOperatorComposition(self, other)
-        else:
-            return OperatorComposition(self, other)
+        elif isinstance(other, Operator):
+            # cast due to https://github.com/python/mypy/issues/16335
+            return OperatorComposition(self, cast(Operator[*Tin2, tuple[torch.Tensor,]], other))
+        return NotImplemented  # type: ignore[unreachable]
```
Comment (ckolbPTB) [inline src/mrpro/operators/LinearOperator.py:217]: Is this needed for something specific?

File: src/mrpro/operators/LinearOperator.py (hunk 1 above)
Comment (ckolbPTB) [reply src/mrpro/operators/LinearOperator.py:206]: similarly in matmul of the EndomorphOperator

File: src/mrpro/operators/LinearOperator.py [hunk 3]
```diff
@@ -218,7 +268,12 @@ class LinearOperatorElementwiseProductRight(
 
     def adjoint(self, x: torch.Tensor) -> tuple[torch.Tensor,]:
         """Adjoint Operator elementwise multiplication with a tensor."""
-        return self._operator.adjoint(x * self._tensor.conj())
+        conj: complex | torch.Tensor
```
Comment (ckolbPTB) [inline src/mrpro/operators/LinearOperator.py]: `conj = self._scalar.conj() if isinstance(self._scalar, torch.Tensor) else self._scalar.conjugate()` ?

File: src/mrpro/operators/Operator.py [hunk 4]
```diff
@@ -31,21 +34,46 @@ def __matmul__(self: Operator[*Tin, Tout], other: Operator[*Tin2, tuple[*Tin]])
...
+        elif isinstance(other, Operator):
+            return OperatorSum(
+                cast(Operator[*Tin, Tout], other), self
+            )  # cast due to https://github.com/python/mypy/issues/16335
+        return NotImplemented  # type: ignore[unreachable]
+
+    def __mul__(self, other: torch.Tensor | complex) -> Operator[*Tin, Tout]:
         """Operator multiplication with tensor.
 
         Returns lambda x: self(other*x)
```
Comment (ckolbPTB) [inline src/mrpro/operators/Operator.py]: should this actually be `x*other`?

File: src/mrpro/operators/ZeroOp.py [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+        """Apply the operator to the input.
+
+        Parameters
+        ----------
+        x
+            Input tensor.
+
+        Returns
+        -------
+        zeros_like(x)
```
Comment (ckolbPTB) [inline src/mrpro/operators/ZeroOp.py]: or 0 scalar

File: tests/operators/test_zero_op.py [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+    generator = RandomGenerator(seed=0)
+    x = generator.complex64_tensor(2, 3, 4)
+    operator = ZeroOp(keep_shape=True)
+    (actual,) = operator(x)
+    expected = torch.zeros_like(x)
+    torch.testing.assert_close(actual, expected)
+
+
+def test_zero_op_scalar():
+    """Test that the zero operator returns zeros."""
```
Comment (ckolbPTB) [inline tests/operators/test_zero_op.py]: single zero

**Review by ckolbPTB — COMMENTED** [review]
just some minor comments

Some of the doc strings need updating because we do not just support tensors now but also scalar values.

File: src/mrpro/operators/LinearOperator.py (hunk 2 above)
Comment (fzimmermann89) [reply src/mrpro/operators/LinearOperator.py:217]: returning notimplemented is the suggested way to signify, that the magic method does not implement this specific combination and python should try rmatmul on the right hand operator.

File: src/mrpro/operators/LinearOperator.py (hunk 1 above)
Comment (fzimmermann89) [reply src/mrpro/operators/LinearOperator.py:206]: in endomorph operator it helped hiding a mypy buig resulting in a crash, but casting to any afew lines furhter down should also now solve it :D

File: src/mrpro/operators/LinearOperator.py [hunk 7]
```diff
@@ -145,46 +148,93 @@ def __matmul__(
...
+            return LinearOperatorSum(self, other)
+        elif isinstance(other, Operator):
+            # for general operators
+            return OperatorSum(self, other)
+        else:
+            return NotImplemented  # type: ignore[unreachable]
 
-    def __mul__(self, other: torch.Tensor) -> LinearOperator:
+    def __mul__(self, other: torch.Tensor | complex) -> LinearOperator:
         """Operator elementwise left multiplication with tensor.
```
Comment (ckolbPTB) [inline src/mrpro/operators/LinearOperator.py]: multiplication with tensor or scalar.

File: src/mrpro/operators/LinearOperator.py [hunk 8]
```diff
@@ -145,46 +148,93 @@ def __matmul__(
...
+            else:
+                return LinearOperatorElementwiseProductLeft(self, other)
+        elif isinstance(other, torch.Tensor):
+            return LinearOperatorElementwiseProductLeft(self, other)
+        else:
+            return NotImplemented  # type: ignore[unreachable]
 
-    def __rmul__(self, other: torch.Tensor) -> LinearOperator:
+    def __rmul__(self, other: torch.Tensor | complex) -> LinearOperator:
         """Operator elementwise right multiplication with tensor.
```
Comment (ckolbPTB) [inline src/mrpro/operators/LinearOperator.py]: multiplication with tensor or scalar.

File: src/mrpro/operators/LinearOperator.py [hunk 9]
```diff
@@ -218,7 +268,8 @@ class LinearOperatorElementwiseProductRight(
 
     def adjoint(self, x: torch.Tensor) -> tuple[torch.Tensor,]:
         """Adjoint Operator elementwise multiplication with a tensor."""
```
Comment (ckolbPTB) [inline src/mrpro/operators/LinearOperator.py]: tensor or scalar

File: src/mrpro/operators/LinearOperator.py [hunk 10]
```diff
@@ -231,7 +282,8 @@ class LinearOperatorElementwiseProductLeft(
...
```
Comment (ckolbPTB) [inline src/mrpro/operators/LinearOperator.py]: tensor or scalar

File: tests/operators/test_zero_op.py [hunk 11]
```diff
@@ -0,0 +1,81 @@
...
+    generator = RandomGenerator(seed=0)
+    x = generator.complex64_tensor(2, 3, 4)
+    operator = ZeroOp(keep_shape=True)
+    (actual,) = operator(x)
+    expected = torch.zeros_like(x)
+    torch.testing.assert_close(actual, expected)
+
+
+def test_zero_op_scalar():
+    """Test that the zero operator return single zero."""
```
Comment (ckolbPTB) [inline tests/operators/test_zero_op.py]: return**s**

**Review by ckolbPTB — APPROVED** [review]
some doc string stuff

File: src/mrpro/operators/LinearOperator.py (hunk 7 above)
Comment (fzimmermann89) [reply src/mrpro/operators/LinearOperator.py]:
```suggestion
        """Operator elementwise left multiplication with tensor/scalar.
```

File: src/mrpro/operators/LinearOperator.py (hunk 8 above)
Comment (fzimmermann89) [reply src/mrpro/operators/LinearOperator.py]:
```suggestion
        """Operator elementwise right multiplication with tensor/scalar.
```

File: src/mrpro/operators/LinearOperator.py (hunk 9 above)
Comment (fzimmermann89) [reply src/mrpro/operators/LinearOperator.py]:
```suggestion
        """Adjoint Operator elementwise multiplication with a tensor/scalar."""
```

File: src/mrpro/operators/LinearOperator.py (hunk 10 above)
Comment (fzimmermann89) [reply src/mrpro/operators/LinearOperator.py]:
```suggestion
        """Adjoint Operator elementwise multiplication with a tensor/scalar."""
```

File: tests/operators/test_zero_op.py (hunk 11 above)
Comment (fzimmermann89) [reply tests/operators/test_zero_op.py]:
```suggestion
    """Test that the zero operator returns single zero."""
```

**Pull request merged**